Issue#937 - step one: script to create version-specific webinstaller/…

yasserare opened the pull request:
In order to add quick-lint-js to webi, we need a file called releases.js that needs a bunch of informations about each QLJS release.
the > make-webinstaller-manifest.js
create a json file that contains all the informations releases.js will need based on quick-lint-js-template.json.
this version of the script is better than the one I showed you in stream, this version takes in arguments, checks if they are valid and display an error message as appropriate.

strager commented on dist/webinstaller/quick-lint-js-template.json:
```diff
@@ -0,0 +1,103 @@
+{
+  "release": [ 
+      {
+      "name": "linux.tar.gz",
```
Nitpick: Use Prettier to format this file.

strager commented on dist/webinstaller/make-webinstaller-manifest.js:
```diff
@@ -0,0 +1,47 @@
+    }
+  
+    return true;
+  }
+  
+  if (!validateArg(args, '-BaseURI', /^https?:\/\/.*\/$/)) {
```
This indentation is confusing. Use Prettier to reformat this code.

strager commented:
Must fix: See the "check copyright" CI job failures.

yasserare commented on dist/webinstaller/make-webinstaller-manifest.js:
```diff
@@ -0,0 +1,47 @@
+const releasesRegex = /\/(\d+\.\d+\.\d+)\/$/;
+const releasesVersion = BaseURI.match(releasesRegex);
+
+const template = fs.readFileSync('quick-lint-js-template.json', 'utf8');
+
+const data = JSON.parse(template);
+
+// set version and insert download with the approriate value 
+for (const release of data.release) {
+    release.version = releasesVersion[1];
+    release.download = release.download.replace('${version}', releasesVersion[1]);
+    release.date = releaseDate;
+}
+
+// Write the modified data back to a file
+fs.writeFileSync(outputFile, JSON.stringify(data, null, 2), 'utf8');
```
I will try this ! this seems much better than what i wrote.

yasserare commented on dist/webinstaller/make-webinstaller-manifest.js:
```diff
@@ -0,0 +1,47 @@
+  
+
+//get the version
+const releasesRegex = /\/(\d+\.\d+\.\d+)\/$/;
+const releasesVersion = BaseURI.match(releasesRegex);
```
I wanted to emulate the scoop script (make-manifest.go) as much as possible after you pointed me to it, I also assumed that this script would be ran in every release folder by another script that gives it the URL of the folder at that moment.
that seems inaccurate given that we included the flags ( -releaseDate ) and now this.
That seems better so I will do it.

strager replied:
> I wanted to emulate the scoop script (make-manifest.go) as much as possible

Nitpick: dist/scoop/make-manifest.go hard-codes the version number inside the template. (That's not the best idea...) It doesn't parse the URL.

yasserare commented:
in commit #2:
        [X] correct release to relases.
        [X] changed how the script takes input(used slice and saved each input in a variabe).
        [X]  fixed (This webinstaller Markdown link is broken.).
        [X] replace README.md  typos.
        [X] included copyright.
        [X]  simple string replacements.
        [X] use version flag instead of BaseURI.
        [X] formatted the Files using prettier.

strager commented on dist/webinstaller/make-webinstaller-manifest.js:
```diff
@@ -0,0 +1,63 @@
+const fs = require("fs");
+
+const args = process.argv.slice(2);
+const releaseVersion = process.argv[3];
+const releaseDate = process.argv[5];
+const outputFile = process.argv[7];
+
+if (!args.includes("-releaseVersion")) {
+  console.log("error: missing -releaseVersion");
+  return 0;
+}
+if (!args.includes("-releaseDate")) {
+  console.log("error: missing -releaseDate");
+  return 0;
+}
+if (!args.includes("-Out")) {
+  console.log("error: missing -Out");
+  return 0;
+}
```
Must fix: This code still does the wrong thing if arguments are not given in a particular order. We should allow *any* order.

strager commented on dist/webinstaller/make-webinstaller-manifest.js:
```diff
@@ -0,0 +1,63 @@
+    "error: invalid -releaseVersion; must match regular expression: " +
+      releaseVersionRegex
+  );
+  return 0;
```
Must fix: We should `process.exit(1);` or something instead. Failures should be signaled with a non-zero process exit code.

strager commented on dist/webinstaller/make-webinstaller-manifest.js:
```diff
@@ -0,0 +1,63 @@
+}
+
+let data = fs.readFileSync("quick-lint-js-template.json", "utf8");
+data = data.replace(/{version}/g, releaseVersion);
+data = data.replace(/\${releaseDate}/g, releaseDate);
```
Why do we use `{}` in one case and `${}` in another case? We should be consistent.

strager reviewed (changes requested):
This is looking much better than before. :+1:

yasserare commented:
in commit#3:
- Made the code take the arguments in no particular order.
- Replaced `retrun(0)` with `process.exit(1)` in failures.
-  Replaced `/{version}/g` with `/\${version}/g` for consistency.

strager commented on dist/webinstaller/quick-lint-js-template.json:
```diff
@@ -0,0 +1,103 @@
+      "channel": "stable",
+      "date": "${releaseDate}",
+      "os": "linux",
+      "arch": "",
```
Is an empty arch allowed? This is an x86_64/amd64 package.

strager replied:
> I think it would be better if we add the arch: x86_64/amd64 to those empty values.

Yes, please do this.

yasserare replied:
Done.

strager reviewed (approved):
The installer script looks good to me. Let me know when I should land your patch.

yasserare commented:
> The installer script looks good to me. Let me know when I should land your patch.

Land it ASAP if you can, so we can finish off this issue "https://github.com/quick-lint/quick-lint-js/issues/937"

strager commented on dist/webinstaller/quick-lint-js-template.json:
```diff
@@ -7,7 +7,7 @@
       "date": "${releaseDate}",
       "os": "linux",
-      "arch": "",
+      "arch": "x86_64/amd64",
```
Must fix: This doesn't look right. I would expect either "x86_64" or "amd64", not both. Did you test this? What does webinstaller expect? What do other webinstaller packages use?

yasserare replied:
You are right. We can't really do that, I will create an object for each of the 3 architectures for the 3 OS's.

yasserare replied:
> You are right. We can't really do that, I will create an object for each of the 3 architectures for the 3 OS's.

@strager  does this seems like the right thing to do ?

yasserare commented:
- Added the appropriate architectures (aarch64 for arm64), (armv7l for armhf and arm )...
- Webi extensions don't recognize (tar.gz) so I used (tar) which is what was used by other packages.
- Fixed the formatting
I tested on linux mint and windows and it worked with success.

strager reviewed (changes requested):
Looks good to me.
